File: ui/core/wasm/bulk_sim/merge.ts
import { BulkGearResult, DistributionMetrics } from '../../proto/api';
import { ConcurrentBulkSimCandidateResult } from './types';

export const hasBulkSimStageError = (baseline: ConcurrentBulkSimCandidateResult | undefined, results: ConcurrentBulkSimCandidateResult[]): boolean => {
	return !!baseline?.error || results.some(result => !!result.error);
};

export const cleanBulkSimDpsMetrics = (metrics: DistributionMetrics | undefined): DistributionMetrics | undefined => {
	if (!metrics) return undefined;
	// allValues is kept: selectBulkSimSurvivors needs the per-iteration values. It is
	// stripped again in bulkSimCandidateResultToProto, so it never reaches the frontend.
	// Mutating in place is safe - the input is a freshly decoded worker result with no
	// other referent - and skips cloning the allValues array (one float per iteration).
	metrics.hist = [];
	return metrics;
};

export const mergeBulkSimCandidateResults = (
	result: ConcurrentBulkSimCandidateResult | undefined,
	additionalResult: ConcurrentBulkSimCandidateResult,
): ConcurrentBulkSimCandidateResult => {
	// Nothing carried over: the additional run is the whole result.
	if (!result) return additionalResult;
	if (result.error) return result;
	if (additionalResult.error) return additionalResult;

	return {
		candidate: result.candidate,
		dpsMetrics: mergeBulkSimDistributionMetrics(result.dpsMetrics, additionalResult.dpsMetrics),
	};
};

const mergeBulkSimDistributionMetrics = (
	metrics: DistributionMetrics | undefined,
	additionalMetrics: DistributionMetrics | undefined,
): DistributionMetrics | undefined => {
	if (!metrics) return additionalMetrics;
	if (!additionalMetrics) return metrics;

	const metricsAggregator = getBulkSimDistributionMetricsAggregatorData(metrics);
	const additionalAggregator = getBulkSimDistributionMetricsAggregatorData(additionalMetrics);
	const totalN = metricsAggregator.n + additionalAggregator.n;
	if (totalN <= 0) return DistributionMetrics.clone(metrics);

	const merged = DistributionMetrics.create({
		avg: (metrics.avg * metricsAggregator.n + additionalMetrics.avg * additionalAggregator.n) / totalN,
		max: metrics.max,
		maxSeed: metrics.maxSeed,
		min: metrics.min,
		minSeed: metrics.minSeed,
		hist: { ...metrics.hist },
		// concat, not slice-then-push: one entry per iteration means this runs to six
		// figures, so it allocates once at the final size instead of regrowing, and it
		// takes an array rather than spread call args (which throw RangeError past the
		// engine's argument limit).
		allValues: metrics.allValues.concat(additionalMetrics.allValues),
		aggregatorData: {
			n: totalN,
			sumSq: metricsAggregator.sumSq + additionalAggregator.sumSq,
		},
	});

	if (additionalMetrics.max > merged.max) {
		merged.max = additionalMetrics.max;
		merged.maxSeed = additionalMetrics.maxSeed;
	}
	if (additionalMetrics.min == 0 || additionalMetrics.min < merged.min) {
		merged.min = additionalMetrics.min;
		merged.minSeed = additionalMetrics.minSeed;
	} else if (additionalMetrics.min == merged.min) {
		merged.minSeed = additionalMetrics.minSeed;
	}
	for (const [roundedDps, count] of Object.entries(additionalMetrics.hist)) {
		merged.hist[Number(roundedDps)] = (merged.hist[Number(roundedDps)] ?? 0) + count;
	}
	merged.stdev = Math.sqrt(Math.max(0, merged.aggregatorData!.sumSq / totalN - merged.avg * merged.avg));
	return merged;
};

const getBulkSimDistributionMetricsAggregatorData = (metrics: DistributionMetrics): { n: number; sumSq: number } => {
	if (metrics.aggregatorData && metrics.aggregatorData.n > 0) return metrics.aggregatorData;
	// Fabricating n=1 makes a merge silently mis-weight the two runs, so surface it
	// instead of letting a wrong average through unnoticed.
	console.warn('[Bulk Sim] distribution metrics are missing aggregator data; merged mean/stdev will be weighted as a single sample');
	const n = 1;
	return {
		n,
		sumSq: (metrics.stdev * metrics.stdev + metrics.avg * metrics.avg) * n,
	};
};

export const bulkSimCandidateResultToProto = (result: ConcurrentBulkSimCandidateResult | undefined): BulkGearResult | undefined => {
	if (!result) return undefined;
	let dpsMetrics = result.dpsMetrics;
	if (dpsMetrics && dpsMetrics.allValues.length > 0) {
		const allValues = dpsMetrics.allValues;
		dpsMetrics.allValues = [];
		const clone = DistributionMetrics.clone(dpsMetrics);
		dpsMetrics.allValues = allValues;
		dpsMetrics = clone;
	}
	return BulkGearResult.create({
		candidateIndex: result.candidate.index,
		gear: result.candidate.gear,
		dpsMetrics,
	});
};

File: sim/core/bulk/merge.go
package bulk

import (
	"log"
	"math"

	"github.com/wowsims/tbc/sim/core"
	"github.com/wowsims/tbc/sim/core/proto"
	googleProto "google.golang.org/protobuf/proto"
)

func mergeBulkSimCandidateResults(result *BulkSimCandidateResult, additionalResult *BulkSimCandidateResult) *BulkSimCandidateResult {
	// Nothing carried over: the additional run is the whole result.
	if result == nil {
		return additionalResult
	}
	if result.Error != nil {
		return result
	}
	if additionalResult == nil || additionalResult.Error != nil {
		return additionalResult
	}

	return &BulkSimCandidateResult{
		Candidate:  result.Candidate,
		DpsMetrics: mergeBulkSimDistributionMetrics(result.DpsMetrics, additionalResult.DpsMetrics),
	}
}

// Combines metrics from two independent sim runs for the same gear set.
// AggregatorData carries the sample count and sum of squares needed to
// recompute the weighted mean/stdev after adaptive extra iterations are
// appended.
func mergeBulkSimDistributionMetrics(metrics *proto.DistributionMetrics, additionalMetrics *proto.DistributionMetrics) *proto.DistributionMetrics {
	if metrics == nil {
		return additionalMetrics
	}
	if additionalMetrics == nil {
		return metrics
	}

	metricsAggregator := bulkSimDistributionMetricsAggregatorData(metrics)
	additionalAggregator := bulkSimDistributionMetricsAggregatorData(additionalMetrics)
	totalN := metricsAggregator.N + additionalAggregator.N
	if totalN <= 0 {
		return googleProto.Clone(metrics).(*proto.DistributionMetrics)
	}

	combinedMetrics := &proto.DistributionMetrics{
		Min:            math.MaxFloat64,
		MinSeed:        math.MaxInt64,
		Hist:           make(map[int32]int32),
		AllValues:      make([]float64, 0),
		AggregatorData: &proto.AggregatorData{},
	}
	core.CombineDistributionMetrics(combinedMetrics, metrics, false, float64(metricsAggregator.N)/float64(totalN))
	core.CombineDistributionMetrics(combinedMetrics, additionalMetrics, true, float64(additionalAggregator.N)/float64(totalN))
	return combinedMetrics
}

func bulkSimDistributionMetricsAggregatorData(metrics *proto.DistributionMetrics) *proto.AggregatorData {
	if metrics.AggregatorData != nil && metrics.AggregatorData.N > 0 {
		return metrics.AggregatorData
	}
	// Fabricating N=1 makes a merge silently mis-weight the two runs, so surface it
	// instead of letting a wrong average through unnoticed.
	log.Printf("[Bulk Sim] WARNING: distribution metrics are missing aggregator data; merged mean/stdev will be weighted as a single sample")
	n := int32(1)
	return &proto.AggregatorData{
		N:     n,
		SumSq: (metrics.Stdev*metrics.Stdev + metrics.Avg*metrics.Avg) * float64(n),
	}
}

func hasBulkSimStageError(baseline *BulkSimCandidateResult, results []*BulkSimCandidateResult) bool {
	if baseline != nil && baseline.Error != nil {
		return true
	}
	for _, result := range results {
		if result != nil && result.Error != nil {
			return true
		}
	}
	return false
}

func bulkSimResultsToCandidates(results []*BulkSimCandidateResult) []BulkSimCandidate {
	return core.MapSlice(results, func(result *BulkSimCandidateResult) BulkSimCandidate {
		return result.Candidate
	})
}

func bulkSimCandidateResultToProto(result *BulkSimCandidateResult) *proto.BulkGearResult {
	if result == nil {
		return nil
	}
	dpsMetrics := result.DpsMetrics
	if dpsMetrics != nil && len(dpsMetrics.AllValues) > 0 {
		dpsMetrics = googleProto.Clone(dpsMetrics).(*proto.DistributionMetrics)
		dpsMetrics.AllValues = nil
	}
	return &proto.BulkGearResult{
		CandidateIndex: result.Candidate.Index,
		Gear:           result.Candidate.Gear,
		DpsMetrics:     dpsMetrics,
	}
}

func cleanBulkSimDpsMetrics(metrics *proto.DistributionMetrics) *proto.DistributionMetrics {
	if metrics == nil {
		return nil
	}
	// AllValues is kept: selectBulkSimSurvivors needs the per-iteration values. It is
	// stripped again in bulkSimCandidateResultToProto, so it never reaches the frontend.
	clone := googleProto.Clone(metrics).(*proto.DistributionMetrics)
	clone.Hist = nil
	return clone
}

func bulkSimAbortedError() *proto.ErrorOutcome {
	return &proto.ErrorOutcome{Type: proto.ErrorOutcomeType_ErrorOutcomeAborted, Message: "Bulk Sim Aborted"}
}
